### PR TITLE
Slow request body copy

### PR DESCRIPTION
**Pull request by arthurdarcet**
When sending a request with a large body (e.g.`session.post('…', data=b'0' * 2_000_000_000)`,) `PayloadWriter._write` is taking very long (5+ seconds for a 2G body).

This is especially problematic because this locks the event loop and nothing else gets executed (so if you are running this in a request handler, the whole server becomes unresponsive).

As far as I understand this, there is no need to actually copy the data before passing it to the underlying asyncio transport: its `write` method handles fragmented chunks and we spare a big `''.join(buffer)`.

This leads to a x2 speed up for the actual write, what ever the actual body size is (I tested with bodies from 2KB to 2GB)

## What do these changes do?

When the asyncio transport is open, leave the buffering to the OS and write data directly to it.
I have also factorised some code, to keep the `_buffer` / `_transport` variables in a clear state (exactly one of them is None at all time)

## Are there changes in behavior for the user?

The headers are not buffered until the first byte of the body is ready if the transport is already there. I'm not sure why this was done, I can't find anything related to that in the git history. @fafhrd91, since this is your code, any idea why sending the headers was buffered even when the transport is ready?

Also, see the changed test: some user might get a ClientPayloadError instead of a ServerDisconnectedError since the headers are now sent even if write_eof raises. This is mostly a side effect of the tests: in real life I would expect users already receiving a bit of both `ClientPayloadError` and `ServerDisconnectedError` for this use case (depending on when exactly the connection was cut off).

### CODE REVIEW

**Comment by arthurdarcet** [comment]
Also, kind of related:
I have not seen any section of the documentation relating to large request/response bodies, both from the server and client point of view.
There are some gotchas when dealing with this in production, to avoid locking the event loop and the server, and it would be nice to gather a list of best practice on how to handle heavy requests.
This section could also mention that doing CPU-intensive jobs on the main thread/process will lock the GIL, the event loop, and the server which might be unexpected for someone coming from multi-threaded frameworks like Flask/…

**Comment by arthurdarcet** [comment]
And lastly, using `session.post('…', data=io.BytesIO(…))` makes everything work correctly, but if I'm not mistaken, this isn't documented anywhere (i had to read the code to see that chunking was happening for BytesIO and not for raw bytes)

I think one of three things should be done:
 - This should be documented somewhere, probably in this same "big requests" section of the docs I mentioned earlier
 - The `payload.BytesPayload` class should do some chunking automatically in my opinion, but I can understand that it makes sense to reserve that for BytesIO payloads.
 - The `BytesPayload` class could log a warning when a payload above ~1MB is passed directly (something like `2 * streams.DEFAULT_LIMIT`?)

Let me know which solution(s) sound good, and I can open PR for them

File: aiohttp/http_writer.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ class PayloadWriter(AbstractPayloadWriter):
 
     def __init__(self, stream, loop, acquire=True):
         self._stream = stream
-        self._transport = None
+        self._transport, self._buffer = None, []
```
Comment (fafhrd91) [inline aiohttp/http_writer.py]: use separate lines

File: aiohttp/http_writer.py [hunk 2]
```diff
@@ -138,29 +138,38 @@ def __init__(self, stream, loop, acquire=True):
         self.output_size = 0
 
         self._eof = False
-        self._buffer = []
         self._compress = None
         self._drain_waiter = None
 
         if self._stream.available:
-            self._transport = self._stream.transport
+            self._transport, self._buffer = self._stream.transport, None
```
Comment (fafhrd91) [inline aiohttp/http_writer.py]: separate lines

File: tests/test_web_server.py [hunk 3]
```diff
@@ -90,8 +90,9 @@ def handler(request):
     server = yield from raw_test_server(handler, logger=logger)
     cli = yield from test_client(server)
 
-    with pytest.raises(client.ServerDisconnectedError):
-        yield from cli.get('/path/to')
+    resp = yield from cli.get('/path/to')
+    with pytest.raises(client.ClientPayloadError):
```
Comment (fafhrd91) [inline tests/test_web_server.py]: why exception is different?

**Comment by fafhrd91** [comment]
regarding payloads. I agree. you are welcome to make PR!

**Comment by arthurdarcet** [comment]
@fafhrd91 you agree with using chunking for `BytesPayload`, or with logging a warning for large `bytes` payloads?

**Comment by fafhrd91** [comment]
all tree. but I am not sure if chunking in BytesPayload makes sense.

**Comment by arthurdarcet** [comment]
i would say, either we handle the chunking in `BytesPayload` and it does not really make sense to log a warning (nothing bad would be happening with big payloads); or we do not do any chunking there (like now), and we log something to warn the user that sending big payloads as bytes directly is a bad idea

For the documentation, I … don't really have the time to write it… but I'm planning on writing a blog article related to our experience with all this. I think it will be easy to extract some guidelines from there and i will open a doc PR when i'm ready. Do you want me to open another pending issue to track this until then?

File: aiohttp/http_writer.py (hunk 1 above)
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: done

File: aiohttp/http_writer.py (hunk 2 above)
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: done

**Comment by fafhrd91** [comment]
we have already doc related issue for payload system.

lets add log to BytesPayload.

File: tests/test_web_server.py (hunk 3 above)
Comment (arthurdarcet) [reply tests/test_web_server.py]: until now, the headers were buffered and not sent to the transport until the first byte of the body was written.

I changed that to send the headers immediately if the transport is ready when `write_headers` is called; it impacts this test because if the server cancels the `write_eof`, the headers are still already sent, so the client does not see the ServerDisconnectedError and instead I have to try to read the body to get a `ClientPayloadError` (the server did not send a complete body)

see the last paragraph of the issue

Comment (fafhrd91) [reply tests/test_web_server.py]: oh, I see. reason is performance for small responses. for small response it usually one syscall, with your change it is two. so I'd like to measure performance on small responses.

Comment (fafhrd91) [reply tests/test_web_server.py]: other note. buffering is needed because of http pipelining. but if we decide on #2109 then it may go away.

**Comment by fafhrd91** [comment]
@asvetlov any comments?

File: aiohttp/web_fileresponse.py [hunk 4]
```diff
@@ -20,17 +20,16 @@
...
 
-    def set_transport(self, transport):
-        self._transport = transport
-
-        if self._drain_waiter is not None:
-            waiter, self._drain_maiter = self._drain_maiter, None
-            if not waiter.done():
-                waiter.set_result(None)
+    def __init__(self, *args, **kwargs):
+        self.__buffer = []
```
Comment (asvetlov) [inline aiohttp/web_fileresponse.py]: It's mostly taste question but why don't invite new name for this buffer instead of mangling?
We never use mangling in aiohttp.

File: aiohttp/http_writer.py [hunk 5]
```diff
@@ -130,6 +130,7 @@ class PayloadWriter(AbstractPayloadWriter):
     def __init__(self, stream, loop, acquire=True):
         self._stream = stream
         self._transport = None
+        self._buffer = []
```
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py:128]: if i remember correctly, aiohttp switched from bytearrays to list for performance reasons too, so this probably need to be benchmarked…

File: aiohttp/http_writer.py [hunk 6]
```diff
@@ -138,29 +139,39 @@ def __init__(self, stream, loop, acquire=True):
...
 
     def set_transport(self, transport):
         self._transport = transport
 
-        chunk = b''.join(self._buffer)
-        if chunk:
-            transport.write(chunk)
-            self._buffer.clear()
+        if self._buffer is not None:
+            for b in self._buffer:
```
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: if i'm understanding this correctly, the buffer is usually just one item, the headers, so sending them in a for loop or joining won't change much.

The tradeoff however is when the buffer is exactly a small chunk (the headers) and one very large chunk (the body). This will copy the body to create a slightly larger chunk and will be much slower than calling write twice.

I have no idea of how the internals of asyncio works, so I'll of course defer to you here ; but do we have benchmarks for this?

File: aiohttp/http_writer.py [hunk 7]
```diff
@@ -138,29 +139,39 @@ def __init__(self, stream, loop, acquire=True):
...
                 waiter.set_result(None)
 
+    @asyncio.coroutine
+    def get_transport(self):
+        if self._transport is None:
+            if self._drain_waiter is None:
+                self._drain_waiter = create_future(self.loop)
+            yield from self._drain_waiter
+
+        assert self._transport is not None
```
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: nothing that i can think of, i was using the asserts as safeguards/way to comment the code. I'll remove them

File: aiohttp/web_fileresponse.py (hunk 4 above)
Comment (arthurdarcet) [reply aiohttp/web_fileresponse.py]: yes, you're perfectly right, i'll update the PR with a new name instead of mangling

File: aiohttp/http_writer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/http_writer.py:128]: We just follow asyncio streams design, it switched from bytearrays to lists and back to bytearrays :)

File: aiohttp/http_writer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/http_writer.py]: Well, nevermind.

[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: i'll try to write some benchmark of how multiple syscalls behave. I'm pretty sure the transport is almost always set before the body is written, so it's not really buffered here.

The main optimisation in all this is sending the buffered data as soon as the transport is set. Right now, the buffered data waits for the next chunk, and then is joined with this new chunk and everything is sent together. That behaves very poorly with a large body because it means copying everything to avoid a second syscall. For very small buffers it probably makes way more sense to bundle everything and do only one syscall

[the remark this answers is not in the document]
Comment (fafhrd91) [reply aiohttp/http_writer.py]: Should we check size of chunk and do join for small chunk?

File: tests/test_web_server.py (hunk 3 above)
Comment (arthurdarcet) [reply tests/test_web_server.py]: See #2179 for some benchmarks. I'm not really sure how to benchmark the syscalls, but impact seems negligible

File: aiohttp/http_writer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py:128]: I tried to benchmark bytearray vs join in #2179, but the results were inconclusive at best (or lean in favor of `b''.join`). I'll try to dig up why asyncio changed again to bytearray

File: aiohttp/http_writer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: checking the size of the chunk would mean buffering the headers until the first chunk is written - as is done currently. I'm not a big fan of that because it makes the code really more complicated (you might end up having data buffered *and* a transport already set). Plus I don't think the syscall matters that much

**Comment by arthurdarcet** [comment]
@asvetlov @fafhrd91 Could you have a look at this? Are there still issue blocking this PR?

**Comment by fafhrd91** [comment]
I can not comment on this, I didn't work on aiohttp for long time.  this is question for other contributors.

@asvetlov @kxepal 

btw can we ping contributors somehow?

**Comment by asvetlov** [comment]
I'll try to find a time tomorrow maybe

@aio-libs/aiohttp-committers should work for pinging all aiohttp committers in the list

File: aiohttp/http_writer.py [hunk 8]
```diff
@@ -273,25 +266,19 @@ def write_headers(self, status_line, headers, SEP=': ', END='\r\n'):
...
-    def drain(self, last=False):
+    async def drain(self):
         if self._transport is not None:
-            if self._buffer:
-                self._transport.write(b''.join(self._buffer))
-                if not last:
-                    self._buffer.clear()
-            yield from self._stream.drain()
+            print(self._stream)
+            print(self._stream.drain)
```
Comment (kxepal) [inline aiohttp/http_writer.py]: print

File: aiohttp/http_writer.py [hunk 9]
```diff
@@ -138,29 +139,36 @@ def __init__(self, stream, loop, acquire=True):
         self.output_size = 0
 
         self._eof = False
-        self._buffer = []
         self._compress = None
         self._drain_waiter = None
 
         if self._stream.available:
             self._transport = self._stream.transport
+            self._buffer = None
```
Comment (kxepal) [inline aiohttp/http_writer.py:142]: I think it's not good idea to jungle variable type. How empty buffer `[]` is different from `None`?

Comment (arthurdarcet) [reply aiohttp/http_writer.py:142]: I'm using `[]` and `None` a bit differently here: once the buffer has been emptied into the new transport, it should never be used again. So I would prefer setting it to None and crash if something very wrong happens

File: aiohttp/http_writer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: Benchmarks were added in a separate pull request: https://github.com/aio-libs/aiohttp/pull/2179

File: aiohttp/http_writer.py (hunk 8 above)
Comment (arthurdarcet) [reply aiohttp/http_writer.py]: oops, thanks, fixed

File: aiohttp/http_writer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kxepal) [reply aiohttp/http_writer.py]: Awesome! Thanks!

File: tests/test_client_functional_oldstyle.py [hunk 10]
```diff
@@ -710,7 +710,7 @@ def connection_lost(self, exc):
             await r.read()
             self.assertEqual(1, len(connector._conns))
 
-            with self.assertRaises(aiohttp.ServerDisconnectedError):
+            with self.assertRaises(aiohttp.ClientConnectionError):
```
[the remark this answers is not in the document]
Comment (arthurdarcet) [reply tests/test_client_functional_oldstyle.py]: the headers are now sent even if write_eof raises. This is mostly a side effect of the tests: in real life I would expect users already receiving a bit of both ClientPayloadError and ServerDisconnectedError for this use case (depending on when exactly the connection was cut off).

File: aiohttp/http_writer.py (hunk 9 above)
Comment (kxepal) [reply aiohttp/http_writer.py:142]: Hm...I see. Well, then such crash shold be controlled, imho. Otherwise it would be hard to distinguish transport re-use case from else error.

File: tests/test_client_functional_oldstyle.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kxepal) [reply tests/test_client_functional_oldstyle.py]: Ok, thanks for explanation!

File: aiohttp/http_writer.py (hunk 9 above)
Comment (arthurdarcet) [reply aiohttp/http_writer.py:142]: It's purely internal to PayloadWriter, so everything is controlled where it should be.

This None means that at any given time, exactly one of `(self._transport, self._buffer)` is None. It makes it easier to check the code / prove that it's correct.

**Comment by arthurdarcet** [comment]
I've rebase this against master, again. Could we reach a decision here?

@asvetlov @kxepal

**Comment by asvetlov** [comment]
Looks good.
Could you add `CHANGES/2126.feature` changelog note.

**Comment by arthurdarcet** [comment]
@asvetlov sure, thanks, just pushed the commit. let me know if the message isn't informative enough

**Comment by asvetlov** [comment]
Let's wait for a while for feedback from other devs.

**Comment by asvetlov** [comment]
Looks like nobody objects, let's go ahead.

@arthurdarcet I've dropped `tests/test_client_functional_oldstyle.py` and moved all remaining tests into `tests/test_client_functional.py`.
Could you solve resolved conflicts?
Also 100% test coverage would be awesome.

**Comment by arthurdarcet** [comment]
@asvetlov 

 - rebase is done

 - I pushed two more tests, that cover two branches in `PayloadWriter`. The only things left untested in this class are the tcp cork related functions (i have no idea what that is, so i'm not sure i should try to write the tests...) and an edge case where the PayloadWriter receives two large concurrent writes before the transport is set (i could test this with a manual calls to `PayloadWriter.get_transport`, but that's not really "public-api").

 - I removed a "if" that wasn't covered with https://github.com/aio-libs/aiohttp/pull/2126/commits/78eb7b44f252cf84b4e1a81567df6cd91402225f . I'm pretty confident it was impossible for the if condition to become false, but please review this commit


(Also, sorry I push over you commit fixing my 2126.feature file and I haven't found a way to cherry pick your commit back, so I rewrote your changes in my original commit)

**Comment by asvetlov** [comment]
@arthurdarcet thanks.
For your information: rebase is not required, just merge is enough. The PR is squashed on merging anyway.
